ui/src/hooks/useHidRpc.ts

```diff
@@ -25,6 +26,128 @@ interface sendMessageParams {
+    if (timeSinceConnected > HANDSHAKE_TIMEOUT) {
+      logger.error(`Handshake timed out after ${timeSinceConnected}ms`);
+      return true;
+    }
+
+    return false;
+  }
+
+  const sendHandshake = (initial: boolean) => {
+    if (handshakeCompleted) return;
+
+    attempts++;
+    lastSendTime = new Date();
+
+    if (!initial && shouldGiveUp()) {
+      if (handshakeInterval) {
+        clearInterval(handshakeInterval);
+        handshakeInterval = null;
+      }
+      return;
+    }
+
+    let data: Uint8Array | undefined;
+    try {
+      const message = new HandshakeMessage(HID_RPC_VERSION);
+      data = message.marshal();
+    } catch (e) {
+      logger.error("Failed to marshal message", e);
+      return;
+    }
+    if (!data) return;
+    rpcHidChannel.send(data as unknown as ArrayBuffer);
+
+    if (initial) {
+      handshakeInterval = setInterval(() => {
+        sendHandshake(false);
+      }, 1000);
+    }
+  };
+
+  const onMessage = (ev: MessageEvent) => {
+    const message = unmarshalHidRpcMessage(new Uint8Array(ev.data));
+    if (!message || !(message instanceof HandshakeMessage)) return;
+
+    if (!message.version) {
+      logger.error("Received handshake message without version", message);
+      return;
+    }
+
+    if (message.version > HID_RPC_VERSION) {
+      // we assume that the UI is always using the latest version of the HID RPC protocol
+      // so we can't support this
+      // TODO: use capabilities to determine rather than version number
+      logger.error("Server is using a newer version than the client", message);
+      return;
+    }
+
+    setRpcHidProtocolVersion(message.version);
+
+    // clean up 
+    const timeUsed = lastSendTime ? Date.now() - lastSendTime.getTime() : 0;
+    logger.info(`Handshake completed in ${timeUsed}ms after ${attempts} attempts (Version: ${message.version} / ${HID_RPC_VERSION})`);
+    resetHandshake({ completed: true });
+
+    rpcHidChannel.removeEventListener("message", onMessage);
+  };
+
+  const resetHandshake = ({ lastConnectedTime: newLastConnectedTime, completed }: { lastConnectedTime?: Date | undefined, completed?: boolean }) => {
+    if (newLastConnectedTime) lastConnectedTime = newLastConnectedTime;
+    lastSendTime = undefined;
+    attempts = 0;
+    if (completed !== undefined) handshakeCompleted = completed;
+    if (handshakeInterval) {
+      clearInterval(handshakeInterval);
+      handshakeInterval = null;
+    }
+  };
+
+  const onConnected = () => {
+    resetHandshake({ lastConnectedTime: new Date() });
+    logger.info("Channel connected");
+
+    sendHandshake(true);
+    rpcHidChannel.addEventListener("message", onMessage);
+  };
+
+  const onClose = () => {
+    resetHandshake({ lastConnectedTime: undefined, completed: false });
+
+    logger.info("Channel closed");
+    setRpcHidProtocolVersion(null);
+
+    rpcHidChannel.removeEventListener("message", onMessage);
+  };
+
+  rpcHidChannel.addEventListener("open", onConnected);
```

Comment:
```suggestion
  rpcHidChannel.addEventListener("close", onClose); 

  // handle case where channel is already open when the hook is mounted
  if (rpcHidChannel.readyState === "open") {
    onConnected();
  } else {
    rpcHidChannel.addEventListener("open", onConnected);
  }
```

Reply:
If we only add the even listener if we're not already open then it won't get double-called